Port `min` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61450
* #61449

Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

ysiraichi committed:
Port `min` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -451,15 +486,13 @@ std::tuple<Tensor, Tensor> max(const Tensor& self, int64_t dim, bool keepdim) {
 
 std::tuple<Tensor, Tensor> min(const Tensor& self, int64_t dim, bool keepdim) {
+  // CUDA and CPU dispatch keys are handled by the structured kernel implementation.
+  TORCH_INTERNAL_ASSERT(self.is_quantized());
```
Same as last PR, I think it's fine to remove this (and rename `min()` to imply that it's a quantized kernel)

bdhirsh commented on test/test_autograd.py:
```diff
@@ -7595,6 +7595,7 @@ def fn(v):
             nnz = 0 if empty_nnz else 5
             _test(sparse_size + dense_size, len(sparse_size), nnz, device)
 
+    @skipMeta
```
What causes this to break?

ysiraichi replied:
In fact, there are many more tests that I'm adding `@skipMeta`. That's mainly because of `torch.sparse_coo_tensor`. Since both `max` and `min` are now structured, it ends up bumping into some error when trying to get an accessor for these meta tensors. It fails here:

https://github.com/pytorch/pytorch/blob/2e49c5dc37bce5ce8ba463a59aacb5d3e4a638b6/c10/core/TensorImpl.h#L1094-L1098

bdhirsh replied:
Cool!

I stared at the `torch.sparse_coo_tensor()` code a bit - I think we hit issues because part of the logic for validating a sparse tensor is to make sure that all of the indices are in bounds, which requires reading the data.

I guess there's no reason why we'd expect sparse meta to work out of the box - and it probably needs some design to figure out what that interaction should look like. So seems reasonable to skip these for now.

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented:
The CI failure is not related to this PR.

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741713).

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

bdhirsh committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741713).

bdhirsh commented:
Actually, is that test failure unrelated? It's a test suite for `meta` (`TestSparseMETA`), looks meta tensor related. The failing check lives here: https://github.com/pytorch/pytorch/blob/6c8cb9bd76299be00d787be2ba02daed0b3921a6/c10/core/TensorImpl.h#L1151 - meta tensors have storage (so `has_storage()` returns true), but it's nullptr so `storage_initialized()` returns false).

This sounds like a meta coverage thing where enabling the meta API for `min` may have gotten some tests further along, so you might need to mark a test in `test_sparse.py` with `skip_meta`?

ezyang commented:
Hum, we probably shouldn't be running any of the sparse tests under meta tensors. Oops!

ysiraichi commented:
@bdhirsh Yeah. That does seem to be the case. I thought I had `@skipMeta` all of them, though.
@ezyang Does adding `"meta"` as the `except_for` parameter when instantiating tests for both `TestSparse` and `TestSparseUnaryUfuncs` sounds reasonable?

ezyang commented:
> Does adding "meta" as the except_for parameter when instantiating tests for both TestSparse and TestSparseUnaryUfuncs sounds reasonable?

sgtm

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

ezyang commented:
@bdhirsh is on vacation this week; if you want this imported faster than that give a holler

ysiraichi commented:
Don't worry. I still want to do some refactoring in this PR.

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

bdhirsh committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741713).

bdhirsh commented:
@ysiraichi let me know if you're still doing any refactoring. Otherwise I'll go ahead and try to land!

ysiraichi commented:
@bdhirsh Sorry for the delay. I'm done refactoring this PR.

bdhirsh committed:
Update on "Port `min` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741713](https://our.internmc.facebook.com/intern/diff/D29741713)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741713).
